drivers/adc/adc_b91.c

```diff
@@ -0,0 +1,372 @@
+
+	case ADC_ACQ_TIME_DEFAULT:
+		sample_cycl = ADC_SAMPLE_CYC_6;
+		break;
+
+	default:
+		LOG_ERR("Selected ADC acquisition time is not supported.");
+		return -EINVAL;
+	}
+
+	/* Check for valid pins configuration */
+	input_positive = adc_b91_get_pin(channel_cfg->input_positive);
+	input_negative = adc_b91_get_pin(channel_cfg->input_negative);
+	if ((input_positive == (uint8_t)ADC_VBAT || input_negative == (uint8_t)ADC_VBAT) &&
+		channel_cfg->differential) {
+		LOG_ERR("VBAT pin is not available for differential mode.");
+		return -EINVAL;
+	} else if (channel_cfg->differential && (input_negative == (uint8_t)NOINPUTN)) {
+		LOG_ERR("Negative input is not selected.");
+		return -EINVAL;
+	}
+
+	/* Init ADC */
+	m_data.differential = channel_cfg->differential;
+	vref = DT_STRING_TOKEN(DT_DRV_INST(0), vref);
+	sample_freq = DT_STRING_TOKEN(DT_DRV_INST(0), sample_freq);
+	adc_init(vref, pre_scale, sample_freq);
+	adc_set_vbat_divider(ADC_VBAT_DIV_OFF);
+	adc_set_tsample_cycle(sample_cycl);
+
+	/* Init ADC Pins */
+	if (channel_cfg->differential) {
+		/* Differential pins configuration */
+		adc_pin_config(ADC_GPIO_MODE, input_positive);
+		adc_pin_config(ADC_GPIO_MODE, input_negative);
+		adc_set_diff_input(channel_cfg->input_positive, channel_cfg->input_negative);
+	} else if (input_positive == (uint8_t)ADC_VBAT) {
+		/* Single-ended Vbat pin configuration */
+		adc_set_diff_input(ADC_VBAT, GND);
+	} else {
+		/* Single-ended GPIO pin configuration */
+		adc_pin_config(ADC_GPIO_MODE, input_positive);
+		adc_set_diff_input(channel_cfg->input_positive, GND);
+	}
+
+	return 0;
+}
+
+/* API implementation: read */
+static int adc_b91_read(const struct device *dev,
+			const struct adc_sequence *sequence)
+{
+	int status;
+
+	adc_context_lock(&m_data.ctx, false, NULL);
+	status = adc_b91_adc_start_read(dev, sequence);
+	adc_context_release(&m_data.ctx, status);
+
+	return status;
+}
+
+#ifdef CONFIG_ADC_ASYNC
+/* API implementation: read_async */
+static int adc_b91_read_async(const struct device *dev,
+			      const struct adc_sequence *sequence,
+			      struct k_poll_signal *async)
+{
+	int status;
+
+	adc_context_lock(&m_data.ctx, true, async);
+	status = adc_b91_adc_start_read(dev, sequence);
+	adc_context_release(&m_data.ctx, status);
+
+	return status;
+}
+#endif /* CONFIG_ADC_ASYNC */
+
+static const struct adc_driver_api adc_b91_driver_api = {
+	.channel_setup = adc_b91_channel_setup,
+	.read = adc_b91_read,
+#ifdef CONFIG_ADC_ASYNC
+	.read_async = adc_b91_read_async,
+#endif
+	.ref_internal = DT_STRING_TOKEN(DT_DRV_INST(0), vref) == ADC_VREF_1P2V ?
+			1200 : 900,
+};
+
+#define ADC_INIT(inst)					      \
+	BUILD_ASSERT((inst) == 0,			      \
+		     "multiple instances not supported");     \
+	DEVICE_DT_INST_DEFINE(0,			      \
+			      adc_b91_init, NULL, NULL, NULL, \
+			      POST_KERNEL,		      \
+			      CONFIG_ADC_INIT_PRIORITY,	      \
+			      &adc_b91_driver_api);
+
+DT_INST_FOREACH_STATUS_OKAY(ADC_INIT)
```

Comment:
please don't use `DT_INST_FOREACH_STATUS_OKAY` is multiple instances are not supported. Also try to run `clang-format` on your driver, there seem to be multiple style issues.

Reply:
Done.
I also applied clang-format changes according to your recommendations, before that I just used checkpatch.

Reply:
Reverted back all clang-format changes because it causes a lot of Compliance Checks errors, checkpatch works much more batter in this case.

Reply:
@gmarull 
Checkpatch PASS and Compliance Checks also PASS.
Please approve the PR if you are OK with the driver.